### ISSUE TITLE
Accessibility features for screen readers

### ISSUE THREAD

**Issue by phloxic** [flowplayer/flowplayer#995]
See: https://flowplayer.org/forum/#!/setup:flowplayer-6-screen-reader

Problems:
- testing requires a screen reader installed
- according to experiences with Flowplayer Flash there's basically only one program around - JAWS http://www.freedomscientific.com/Products/Blindness/JAWS - Used to work with IE only, now seems to support Firefox

@nnarhinen - we should come to a decision whether we are _planning_ to implement this or not.

**Comment by deadplant** [comment]
for testing there is a free plugin screen-reader for chome https://chrome.google.com/webstore/detail/chromevox/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en

It should be a relatively simple feature to implement when using the html5 engine.  At a minimum, allowing the user to tab focus through the player controls and providing good alt-text for them should be enough for the screen reader.

It (accessibility including screen reading) is becoming a requirement for more and more governments and organizations.

**Comment by nnarhinen** [comment]
I'm all for it, no timeframe though

**Comment by deadplant** [comment]
Here is some more info that I hope might help put someone on the right track when/if they go to work on this.

There are essentially 3 things we need for the screen reader to work well.

- tab-key navigation of the player controls.   (a tabindex on each relevant div or anchor)
- a title attribute on each of those.   title="full screen" so the reader can speak something sensible
- keyboard activation of the focused player control.  enter/spacebar for most things, perhaps arrow keys for the volume slider and scrub bar.

pitfalls:
- focused elements should get a css outline when they have been given focus by the keyboard (this is the desired behaviour for accessibility)  *but* should not get the outline when given focus by a mouse click.

hopefully that is helpful to somebody at some point.  :)

**Comment by blacktrash** [comment]
I wonder whether something could be done with the builtin keyboard shortcut support: https://flowplayer.org/docs/setup.html#keyboard

**Comment by deadplant** [comment]
tabbing in and through the player controls (in the html5 player engine) simply requires a tabindex property on each control's html element.

As a temporary solution I am adding those properties after the player finishes rendering.
Then, to activate the selected control via spacebar/enter I bind a callback to the element like this:
$('.fp-mute').keypress(function(evt){ if (evt.keyCode == 13 || evt.keyCode == 32) { player.mute(); } });

If the built-in keyboard controls could catch space/enter keystrokes and activate whichever tabindex has focus that would be nice.

**Comment by qualia100** [comment]
The builtin keyboard shortcuts are great; the problem is - you can't access them without a mouse. Some people are unable to use a mouse and don't even have it. This is why it is important to be able to use a tab to 'activate' those controls. Right now, only mouseover (mouseenter, mouseleave) activate but there is a need to add focus to the that group. Of course, I can adjust the code myself; however I would prefer not to as javascript is not my area of expertise.

**Comment by qualia100** [comment]
In order to be 508 compliant, it should be possible to use a player with a keyboard only (no mouse). Currently, it is not possible as users need to 'activate' the player area in order to get keyboard shortcuts (https://flowplayer.org/docs/setup.html#keyboard) and that can't be done without a mouse. New functionality that would allow users to use tab to select player area would be very useful. Ideally, users should be able to tab through all flowplayer's controls and activate them with a keyboard. 
Please notify when this functionality could be added to the player.

**Comment by qualia100** [comment]
Fullscreen keyboard access ('f' when mouse is over player's area) does not work in Edge. Works fine in other browsers.

**Comment by blacktrash** [comment]
Also does not work in Mac OS Safari. Never did.

**Comment by qualia100** [comment]
My understanding is this is planned for 7.1 release. Is there a timeframe for that release? The ability to access and activate player's controls via tabbing is critical for our product.

**Comment by Honzasejrak** [comment]
sirra23

**Comment by webcast33** [comment]
I see this issue is now scheduled for player release 7.3. Do you have an ECD for that release? Thank you.

**Comment by fdambrosio** [comment]
This is an important feature, thanks

**Comment by fdambrosio** [comment]
can you add the aria-label attribute?  
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute
